### PR TITLE
wasm: refactor common codes around configurations

### PR DESCRIPTION
**Pull request by mathetake**
Refactored wasm common codes around configurations by introducing `WasmConfig` class which performs conversion of restriction configs (and other configs in the future). Resolved the comment https://github.com/istio/proxy/pull/3202#discussion_r579505250.

### CODE REVIEW

File: source/extensions/access_loggers/wasm/config.cc [hunk 1]
```diff
@@ -24,16 +24,12 @@ WasmAccessLogFactory::createAccessLogInstance(const Protobuf::Message& proto_con
...
-      config.config().name(), config.config().root_id(), config.config().vm_config().vm_id(),
-      config.config().vm_config().runtime(),
-      Common::Wasm::anyToBytes(config.config().configuration()), config.config().fail_open(),
-      envoy::config::core::v3::TrafficDirection::UNSPECIFIED, context.localInfo(),
+      config.config(), envoy::config::core::v3::TrafficDirection::UNSPECIFIED, context.localInfo(),
       nullptr /* listener_metadata */);
 
-  auto access_log = std::make_shared<WasmAccessLog>(plugin, nullptr, std::move(filter));
+  auto access_log =
+      std::make_shared<WasmAccessLog>(config.config(), plugin, nullptr, std::move(filter));
```
Comment (PiotrSikora) [inline source/extensions/access_loggers/wasm/config.cc:32]: Couldn't `WasmBaseConfig` be a member of `Plugin` instead?

File: source/extensions/common/wasm/base_config.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+#include "include/proxy-wasm/wasm.h"
+#include "include/proxy-wasm/wasm_vm.h"
+#include "include/proxy-wasm/word.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Wasm {
+
+class WasmBaseConfig {
```
Comment (PiotrSikora) [inline source/extensions/common/wasm/base_config.h:19]: We always use `Base` to denote Proxy-Wasm class, i.e. `PluginBase` is defined in Proxy-Wasm, and `Plugin` is its subclass defined in Envoy, so this should be called `WasmConfig`, or perhaps `PluginConfig`?

File: source/extensions/access_loggers/wasm/config.cc (hunk 1 above)
Comment (mathetake) [reply source/extensions/access_loggers/wasm/config.cc:32]: nice, that would be better. Will do.

**Comment by PiotrSikora** [comment]
@mathetake it looks that you forgot to add some files.

**Comment by mathetake** [comment]
sorry, I missed them. I just Added.

**Comment by PiotrSikora** [comment]
@mathetake that failure is deterministic (coverage is too low) and I believe unrelated to this PR. Merging `main` might fix it.

**Comment by mathetake** [comment]
> Merging main might fix it.
Unfortunately the branch is already up-to-date, so I believe something is broken on main branch, and that's unrelated to this PR since the same happens on other PRs.

**Review by PiotrSikora — APPROVED** [review]
LGTM. Thanks!

**Comment by mathetake** [comment]
coverage failure will be fixed in https://github.com/envoyproxy/envoy/pull/15238

File: source/common/protobuf/utility.h [hunk 3]
```diff
@@ -329,6 +329,26 @@ class MessageUtil {
    */
   static void unpackTo(const ProtobufWkt::Any& any_message, Protobuf::Message& message);
 
+  /**
+   * Convert from google.protobuf.Any to bytes as std::string
+   * @param message source google.protobuf.Any message.
+   *
+   * @return std::string consists of bytes in the input message.
+   */
+  static std::string anyToBytes(const ProtobufWkt::Any& any) {
```
Comment (antoniovicente) [inline source/common/protobuf/utility.h:338]: Would it be possible to add some basic small tests that verify that this method works as intended?

Comment (mathetake) [reply source/common/protobuf/utility.h:338]: Thanks, will do!

Comment (mathetake) [reply source/common/protobuf/utility.h:338]: see https://github.com/envoyproxy/envoy/pull/15326

**Pull request merged**